packages/react/src/components/createValueTrackedComponent.tsx

```diff
@@ -0,0 +1,133 @@
+import React, { createElement } from 'react';
+
+import {
+  attachProps,
+  createForwardRef,
+  dashToPascalCase,
+  isCoveredByReact,
+  mergeRefs,
+} from './react-component-lib/utils';
+
+export interface HTMLStencilElement extends HTMLElement {
+  componentOnReady(): Promise<this>;
+}
+
+interface StencilReactInternalProps<ElementType> extends React.HTMLAttributes<ElementType> {
+  forwardedRef: React.RefObject<ElementType>;
+  ref?: React.Ref<any>;
+}
+
+export const createValueTrackedComponent = <
+  PropType,
+  ElementType extends HTMLStencilElement,
+  ContextStateType = {},
+  ExpandedPropsTypes = {}
+>(
+  tagName: string,
+  ReactComponentContext?: React.Context<ContextStateType>,
+  manipulatePropsFunction?: (
+    originalProps: StencilReactInternalProps<ElementType>,
+    propsToPass: any,
+  ) => ExpandedPropsTypes,
+  defineCustomElement?: () => void,
+) => {
+  if (defineCustomElement !== undefined) {
+    defineCustomElement();
+  }
+
+  const displayName = dashToPascalCase(tagName);
+  const ReactComponent = class extends React.Component<StencilReactInternalProps<ElementType>> {
+    componentEl!: ElementType;
+
+    setComponentElRef = (element: ElementType) => {
+      this.componentEl = element;
+    };
+
+    updateValueIfChanged = (ev: any) => {
```

Comment:
Rather than implement this at the React level, have we considered having `ion-input` wait for `ionChange` to fire before `ionInput` is fired? Something like this:

```diff
private onInput = (ev: Event) => {
    const input = ev.target as HTMLInputElement | null;
    if (input) {
      this.value = input.value || '';

+    await waitForIonChangeToFire();
    }
    this.ionInput.emit(ev as InputEvent);
}
```

This ensures that the value is committed to `ion-input` before any formatting is done. It also ensures that `ionChange` fires before `ionInput`, which is aligned with native `inputs` where `change` fires before `input`. Currently, `ionInput` fires before `ionChange`.

Eric also noted that this impacts Ionic Angular, so implementing this in core should fix the issue there too: https://github.com/ionic-team/ionic-framework/issues/22868#issuecomment-897059622

Reply:
I like the idea of trying to address this problem across both, my concern in doing so was changing existing behavior/causing a breaking change. 

If we feel that event ordering wouldn't meet that criteria, I can investigate this approach more!